### PR TITLE
[component: agw] Session Release & UE Deregister

### PR DESCRIPTION
**Pull request by panyogesh**
Fixes included :
  1. PDU Session Release
  2. UE Deregistration

Testcases :
  1. Unit Testcases of ngap
  2. Testing of PDU Session Release on UERANSIM
  3. Testing of UE Deregistration on UERANSIM

Signed-off-by: YOGESH PANDEY <yogesh@wavelabs.ai>

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/c/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp [hunk 1]
```diff
@@ -171,15 +271,43 @@ int pdu_session_resource_release_request(
...
+      LOG_AMF_APP,
+      //"Converting pdu_session_resource_release_command_transfer to
+      // bstring\n");
+      "filling pdu_session_resource_release_command_transfer\n");
+  // ngap_pdu_ses_release_req->pduSessionResourceToRelReqList.item[0]
+  //    .PDU_Session_Resource_TO_Release_Command_Transfer = blk2bstr(
+  //    &amf_pdu_ses_rel_transfer_req,
+  //    sizeof(
+  //        pdu_session_resource_release_command_transfer));  //
+  //        pdu_res_set_change
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp]: can you remove commented out code?

File: lte/gateway/c/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp [hunk 2]
```diff
@@ -171,15 +271,43 @@ int pdu_session_resource_release_request(
...
+  if (bytes > 0) {
+    OAILOG_INFO(LOG_AMF_APP, "NAS encode success for PDU Release Command\n");
+    //    buffer->slen = bytes;
+    buffer->slen = bytes;
+    for (int i = 0; i < blength(buffer); i++) {
+      OAILOG_INFO(LOG_AMF_APP, "buffer[%d]: %x", i, buffer->data[i]);
+    }
+    //    bytes + 3;  // TODO fix length in all other DownlinkNASTransport
+    // amf_app_def_as.amf_app_handle_nas_dl_req(ue_id, buffer, rc);
+    //    amf_app_handle_nas_dl_req(ue_id, buffer, rc);
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp]: remove commented out code pieces.

File: lte/gateway/c/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp [hunk 3]
```diff
@@ -171,15 +271,43 @@ int pdu_session_resource_release_request(
...
+  //        pdu_session_resource_release_command_transfer));  //
+  //        pdu_res_set_change
   ngap_pdu_ses_release_req->pduSessionResourceToRelReqList.item[0]
       .PDU_Session_Resource_TO_Release_Command_Transfer =
       amf_pdu_ses_rel_transfer_req;
-
   // Send message to NGAP task
+  if (bytes > 0) {
+    OAILOG_INFO(LOG_AMF_APP, "NAS encode success for PDU Release Command\n");
+    //    buffer->slen = bytes;
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp:279]: remove commented out code pieces.

File: lte/gateway/c/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp [hunk 4]
```diff
@@ -171,15 +271,43 @@ int pdu_session_resource_release_request(
...
+    // amf_app_def_as.amf_app_handle_nas_dl_req(ue_id, buffer, rc);
+    //    amf_app_handle_nas_dl_req(ue_id, buffer, rc);
+    ngap_pdu_ses_release_req->nas_msg = bstrcpy(buffer);
+    bdestroy(buffer);
+  } else {
+    bdestroy(buffer);
+    OAILOG_INFO(LOG_AMF_APP, "NAS encode failed for PDU Release Command\n");
+    // OAILOG_ERROR(
+    // LOG_NAS, "Fire error return, Encode DLNASTransport Failed\n",
+    // return rc;
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp]: remove

File: lte/gateway/c/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp [hunk 5]
```diff
@@ -171,15 +271,43 @@ int pdu_session_resource_release_request(
...
+    for (int i = 0; i < blength(buffer); i++) {
+      OAILOG_INFO(LOG_AMF_APP, "buffer[%d]: %x", i, buffer->data[i]);
+    }
+    //    bytes + 3;  // TODO fix length in all other DownlinkNASTransport
+    // amf_app_def_as.amf_app_handle_nas_dl_req(ue_id, buffer, rc);
+    //    amf_app_handle_nas_dl_req(ue_id, buffer, rc);
+    ngap_pdu_ses_release_req->nas_msg = bstrcpy(buffer);
+    bdestroy(buffer);
+  } else {
+    bdestroy(buffer);
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp:288]: if there will be no return in this clause `bdestroy` on lines 302 and 304 can be moved out to if/else clause. However, if NAS encode has failed, do we still want to send msg and return RETURNok?

Comment (ulaskozat) [reply lte/gateway/c/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp:288]: if returning in else clause, we need to free the allocated message.

**Review by ulaskozat — APPROVED** [review]

**Review by ulaskozat — CHANGES_REQUESTED** [review]
update session_manager.pb.go

**Comment by panyogesh** [comment]
Test-Logs-May-26

**Comment by sureshgorijavolu** [comment]
code is merged through #7092

**Pull request merged**